src/core.js

```diff
@@ -221,9 +224,15 @@ const alignWithDOM = function(nodeName, key) {
 
   // Check to see if the node has moved within the parent.
   if (key) {
-    node = keyMap[key];
-    if (node && process.env.NODE_ENV !== 'production') {
-      assertKeyedTagMatches(getData(node).nodeName, nodeName, key);
+    const keyNode = keyMap[key];
+    if (matches(nodeName, key, keyNode)) {
+      node = keyNode;
+    } else if (keyNode) {
```

Comment:
Should move the matches check in here, remove the `matchNode` check in `matches`, put back the currentNode check at the top